detox/src/Detox.js

```diff
@@ -32,75 +32,44 @@ class Detox {
     );
 
     const {artifactsConfig, deviceConfig, session} = config;
+    this._artifactsConfig = artifactsConfig;
     this._deviceConfig = deviceConfig;
     this._userSession = deviceConfig.session || session;
     this._client = null;
     this._server = null;
-    this._artifactsManager = new ArtifactsManager(artifactsConfig);
+    this._artifactsManager = null;
+    this._isInitialized = false;
+    this._initPromise = null;
+    this._rejectInit = _.noop;
 
     this.device = null;
   }
 
-  async init(userParams) {
-    const sessionConfig = await this._getSessionConfig();
-    const params = {
-      launchApp: true,
-      initGlobals: true,
-      ...userParams,
-    };
-
-    if (!this._userSession) {
-      this._server = new DetoxServer({
-        log: logger,
-        port: new URL(sessionConfig.server).port,
-      });
+  init(userParams) {
+    if (!this._initPromise) {
```

Comment:
I just think that it is common sense to abort `detox.init()` promise if it is not done, but we are calling `detox.cleanup()`. Why should this promise be unresolved, if we can reject it? :man_shrugging:

Reply:
I understand and like this approach. Nevertheless we now have in this class these 3 members, all related to init:
 - `_isInitialized`
 - `_initPromise`
 - `_rejectInit`
I can't help but think we ought to replace those with a single ref to an external init controller of some sort, to which all of this be delegated, managing and holding all necessary flags and init state indicators. The code here is already entangled as it is.